### PR TITLE
TensorFlow 中的 RNN 串流

### LINKED ISSUE #4561: Streaming RNNs in TensorFlow
* 原文链接：[Streaming RNNs in TensorFlow](https://hacks.mozilla.org/2018/09/speech-recognition-deepspeech/)
  * Markdown文件：[文件地址](https://github.com/xitu/gold-miner/blob/master/TODO1/speech-recognition-deepspeech.md)
  * PR 地址：
  * 文章分类：**AI**
  * **注意：文件位置在 `TODO1` 文件夹中，不是之前的 `TODO` 了。**
  ----
  * 翻译时间：3 天
  * 校对时间：1 天
  * 翻译积分：3-4 分
  * 校对积分：1-2 分
  * [积分有什么用](https://github.com/xitu/gold-miner/wiki)

### PR DESCRIPTION
**Pull request by sisibeloved**
译文翻译完成，resolve #4561

### CODE REVIEW

**Comment by leviding** [comment]
@sisibeloved 以英文原文为准

**Comment by lsvih** [comment]
认领翻译

File: TODO1/speech-recognition-deepspeech.md [hunk 1]
```diff
@@ -2,42 +2,42 @@
...
 > * 校对者：
 
-# Streaming RNNs in TensorFlow
+# TensorFlow 中的 RNN 串流
 
-The Machine Learning team at [Mozilla Research](http://research.mozilla.org/) continues to work on an automatic speech recognition engine as part of [Project DeepSpeech](https://github.com/mozilla/DeepSpeech), which aims to make speech technologies and trained models openly available to developers. We’re hard at work improving performance and ease-of-use for our open source speech-to-text engine. The upcoming 0.2 release will include a much-requested feature: the ability to do speech recognition live, as the audio is being recorded. This blog post describes how we changed the STT engine’s architecture to allow for this, achieving real-time transcription performance. Soon, you’ll be able to transcribe audio at least as fast as it’s coming in.
+[谋智（Mozilla）研究所](http://research.mozilla.org/)的机器学习团队正在开发一个自动语音识别引擎，它将作为[深度语音（DeepSpeech）项目](https://github.com/mozilla/DeepSpeech)的一部分，致力于向开发人员开放语音识别技术和受训模型。我们正在努力提高我们开源的语音转文本引擎的性能和易用性。即将发布的 0.2 版本将包括一个大家期待已久的特性：在录制音频时实时进行语音识别的能力。这篇博客文章描述了我们是怎样修改 STT（即 speech-to-text，语音转文字）引擎的架构，来达到实现实时转录的性能要求。很快，等到正式版本发布，你就可以体验这一音频转换的功能。
```
Comment (lsvih) [inline TODO1/speech-recognition-deepspeech.md]: 受训模型 -> 预训练模型

Comment (lsvih) [inline TODO1/speech-recognition-deepspeech.md]: 很快，等到正式版本发布 -> 不久之后，当正式版本发布时
（此修改只是为了稍微流畅一点）

File: TODO1/speech-recognition-deepspeech.md [hunk 2]
```diff
@@ -2,42 +2,42 @@
...
 
-When applying neural networks to sequential data like audio or text, it’s important to capture patterns that emerge over time. Recurrent neural networks (RNNs) are neural networks that “remember” — they take as input not just the next element in the data, but also a state that evolves over time, and use this state to capture time-dependent patterns. Sometimes, you may want to capture patterns that depend on future data as well. One of the ways to solve this is by using two RNNs, one that goes forward in time and one that goes backward, starting from the last element in the data and going to the first element. You can learn more about RNNs (and about the specific type of RNN used in DeepSpeech) in [this article by Chris Olah](https://colah.github.io/posts/2015-08-Understanding-LSTMs/).
+当将神经网络应用到诸如音频或文本的顺序数据时，获取数据随着时间推移而出现的模式是很重要的。循环神经网络（RNN）是具有『记忆』的神经网络 —— 它们不仅将数据中的下一个元素作为输入，而且还将随时间演进的状态作为输入，并使用这个状态来获取与时间相关的模式。有时，你可能希望获取依赖未来数据的模式。解决这个问题的方法之一是使用两个 RNN，一个在时序上向前，而另一个向后，即从数据中的最后一个元素开始，到第一个元素。你可以在 [Chris Olah 的这篇文章](https://colah.github.io/posts/2015-08-Understanding-LSTMs/)中了解更多关于 RNN（以及关于 DeepSpeech 中使用的特定类型的 RNN）的知识。
```
Comment (lsvih) [inline TODO1/speech-recognition-deepspeech.md]: 当将 -> 将

Comment (lsvih) [inline TODO1/speech-recognition-deepspeech.md]: 获取 -> 捕获

Comment (lsvih) [inline TODO1/speech-recognition-deepspeech.md]: 而另一个向后，即从数据中的最后一个元素开始，到第一个元素
->
而另一个按向后的时序（即从数据中的最后一个元素开始，到第一个元素）

File: TODO1/speech-recognition-deepspeech.md [hunk 3]
```diff
@@ -2,42 +2,42 @@
...
+DeepSpeech 的当前版本（[之前在 Hacks 上讨论过](https://hacks.mozilla.org/2017/11/a-journey-to-10-word-error-rate/)）使用了用 [TensorFlow](https://www.tensorflow.org/) 实现的双向 RNN，这意味着它需要在开始工作之前具有整个可用的输入。一种改善这种情况的方法是通过实现流式模型：在数据到达时以块为单位进行工作，这样当输入结束时，模型已经在处理它，并且可以更快地给出结果。你也可以尝试在输入中途查看部分结果。
 
 ![This animation shows how the data flows through the network. Data flows from the audio input to feature computation, through three fully connected layers. Then it goes through a bidirectional RNN layer, and finally through a final fully connected layer, where a prediction is made for a single time step.](https://2r4s9p1yi1fa2jd7j43zph8r-wpengine.netdna-ssl.com/files/2018/09/bidirectional.gif)
 
-This animation shows how the data flows through the network. Data flows from the audio input to feature computation, through three fully connected layers. Then it goes through a bidirectional RNN layer, and finally through a final fully connected layer, where a prediction is made for a single time step.
+> 这个动画展示了数据如何在网络间流动。数据通过三个全连接层，从音频输入转变成特征计算。然后通过了一个双向 RNN 层，最后通过最终的对单个时间步长进行预测的全连接层。
```
Comment (lsvih) [inline TODO1/speech-recognition-deepspeech.md]: 最后通过最终的 -> 最后通过

File: TODO1/speech-recognition-deepspeech.md [hunk 4]
```diff
@@ -2,42 +2,42 @@
...
 
-In order to do this, you need to have a model that lets you do the work in chunks. Here’s the diagram of the current model, showing how data flows through it.
+为了做到这一点，你需要有一个分块模型。这是当前模型的图表，显示数据如何流过它。
```
Comment (lsvih) [inline TODO1/speech-recognition-deepspeech.md]: 分块模型 -> 可以分块处理数据的模型

File: TODO1/speech-recognition-deepspeech.md [hunk 5]
```diff
@@ -2,42 +2,42 @@
...
+可以看到，在双向 RNN 中，倒数第二步的计算需要最后一步的数据，倒数第三步的计算需要倒数第二步的数据……如此循环往复。这些是图中从右到左的红色箭头。
 
-We could implement partial streaming in this model by doing the computation up to layer three as the data is fed in. The problem with this approach is that it wouldn’t gain us much in terms of latency: Layers four and five are responsible for almost half of the computational cost of the model.
+通过在数据被馈入时进行到第三层的计算，我们可以实现部分流式处理。这种方法的问题是它在延迟方面不会给我们带来太多好处：第四层和第五层占用了整个模型几乎一半的计算成本。
 
-## Using a unidirectional RNN for streaming
+## 使用单向 RNN 处理串流
 
-Instead, we can replace the bidirectional layer with a unidirectional layer, which does not have a dependency on future time steps. That lets us do the computation all the way to the final layer as soon as we have enough audio input.
+因此，我们可以用单向层替换双向层，单向层不依赖于将来的时间步骤。只要我们有足够的音频输入，就能一直计算到最后一层。
```
Comment (lsvih) [inline TODO1/speech-recognition-deepspeech.md]: 时间步骤 -> 时间步

File: TODO1/speech-recognition-deepspeech.md [hunk 6]
```diff
@@ -103,11 +103,11 @@ def create_inference_graph(batch_size=1, n_steps=16, n_features=26, width=64):
     }
 ```
 
-The graph created by the code above has two inputs and two outputs. The inputs are the sequences and their lengths. The outputs are the logits and a special “initialize_state” node that needs to be run at the beginning of a new sequence. When freezing the graph, make sure you don’t freeze the state variables previous_state_h and previous_state_c.
+上述代码创建的图有两个输入和两个输出。输入是序列及其长度。输出是 `logit` 和一个需要在一个新序列开始运行的特殊节点 `initialize_state`。当固化图像时，请确保不固化状态变量 `previous_state_h` 和 `previous_state_c`。
 
-Here’s code for freezing the graph:
+下面是固化图像的代码:
```
Comment (lsvih) [inline TODO1/speech-recognition-deepspeech.md]: 图像 -> 图

File: TODO1/speech-recognition-deepspeech.md [hunk 7]
```diff
@@ -122,29 +122,29 @@ freeze_graph.freeze_graph_with_def_protos(
...
-4.  Repeat 2 and 3 until data is over.
+1.  运行 `initialize_state` 节点。
+2.  积累音频样本，直到数据足以供给模型（我们使用的是 16 个时间步长，或 320ms）
+3.  将数据供给模型，在某个地方积累输出。
+4.  重复第二步和第三步直到数据结束。
 
-It wouldn’t make sense to drown readers with hundreds of lines of the client-side code here, but if you’re interested, it’s all MPL 2.0 licensed and available on [GitHub](https://github.com/mozilla/DeepSpeech). We actually have two different implementations, [one in Python](https://github.com/mozilla/DeepSpeech/blob/bb299dc26554b2fbf864b7f0115b4baece15bda5/evaluate.py#L233) that we use for generating test reports, and [one in C++](https://github.com/mozilla/DeepSpeech/blob/6f27928841c2595c8dd9d08f482c95ca9e42f4b5/native_client/deepspeech.cc) which is behind our official client API.
+把几百行的客户端代码扔给读者是没有意义的，但是如果你感兴趣的话，这些 [GitHub](https://github.com/mozilla/DeepSpeech) 上的代码都是满足 MPL 2.0 许可的。事实上，我们有两种不同语言的实现，一个用 [Python](https://github.com/mozilla/DeepSpeech/blob/bb299dc26554b2fbf864b7f0115b4baece15bda5/evaluate.py#L233),用来生成测试报告；另一个用 [C++](https://github.com/mozilla/DeepSpeech/blob/6f27928841c2595c8dd9d08f482c95ca9e42f4b5/native_client/deepspeech.cc)，这是我们官方的客户端 API。
```
Comment (lsvih) [inline TODO1/speech-recognition-deepspeech.md]: 这些 [GitHub](https://github.com/mozilla/DeepSpeech) 上的代码都是满足 MPL 2.0 许可的
->
可以查阅 [GitHub 中的代码](https://github.com/mozilla/DeepSpeech)，这些代码均遵循 MPL 2.0 协议

Comment (lsvih) [inline TODO1/speech-recognition-deepspeech.md]: ,用来生成 -> ，用来生成

**Comment by leviding** [comment]
@sisibeloved 可以修改啦

**Comment by leviding** [comment]
@sisibeloved 修改一下吧

**Comment by sisibeloved** [comment]
@leviding 已发布：[https://juejin.im/post/5bcb2975f265da0a8d36c7d8](https://juejin.im/post/5bcb2975f265da0a8d36c7d8)
